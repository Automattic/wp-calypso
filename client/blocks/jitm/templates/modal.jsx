import { Button, Guide } from '@wordpress/components';
import { take } from 'lodash-es/array';
import { useState } from 'react';
import './modal-style.scss';
import Visual from './plans-visual.svg';

export default function ModalTemplate( {
	featureClass,
	CTA,
	trackImpression,
	message,
	description,
	onClick,
	onDismiss,
	title,
	disclaimer,
} ) {
	trackImpression && trackImpression();

	// technically, non-dismissable jitms are authorable, however, that doesn't make any sense as a modal.
	const [ isDismissed, setDismissed ] = useState( [] );

	return isDismissed.includes( featureClass ) ? null : (
		<Guide
			className="modal__main"
			contentLabel={ message }
			onFinish={ () => {
				onDismiss();
				setDismissed( isDismissed.concat( [ featureClass ] ) );
			} }
			pages={ [
				{
					content: (
						<>
							<div className="modal__container">
								{ /* todo: allow specifying this text via jitm configuration */ }
								<p className="modal__limited-offer">{ title }</p>
								<h1 className="modal__heading">{ message }</h1>
								<p className="modal__text">
									{ description }
									<br />
									<Button
										href={ CTA.link }
										isPrimary={ true }
										onClick={ () => {
											onClick();
											setDismissed( isDismissed.concat( [ featureClass ] ) );
										} }
									>
										{ CTA.message }
									</Button>
								</p>
								{ take( disclaimer, 2 ).map( ( line ) => (
									<p className="modal__disclaimer">{ line }</p>
								) ) }
							</div>
							<div className="modal__sidebar">
								<img
									src={ Visual }
									alt="list of available plans â€” premium, business, and ecommerce plans"
								/>
							</div>
						</>
					),
				},
			] }
		/>
	);
}
