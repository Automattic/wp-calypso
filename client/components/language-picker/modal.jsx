/**
 * External dependencies
 */
import PropTypes from 'prop-types';

import React, { PureComponent } from 'react';
import classNames from 'classnames';
import { localize } from 'i18n-calypso';
import { includes, map, noop, partial } from 'lodash';

/**
 * Internal dependencies
 */
import Dialog from 'components/dialog';
import SectionNav from 'components/section-nav';
import SectionNavTabs from 'components/section-nav/tabs';
import SectionNavTabItem from 'components/section-nav/item';
import Search from 'components/search';

class LanguagePickerModal extends PureComponent {
	static propTypes = {
		onSelected: PropTypes.func,
		onClose: PropTypes.func,
	}

	static defaultProps = {
		onSelected: noop,
		onClose: noop,
	}

	constructor( props ) {
		super( props );

		this.state = {
			filter: 'popular',
			search: false,
			selectedLanguageSlug: this.props.selected,
		};
	}

	componentWillReceiveProps( nextProps ) {
		if ( nextProps.selected !== this.state.selectedLanguageSlug ) {
			this.setState( {
				selectedLanguageSlug: nextProps.selected
			} );
		}
	}

	getFilterLabel( filter ) {
		const { translate } = this.props;

		switch ( filter ) {
			case 'popular':
				return translate( 'Popular languages', { textOnly: true } );
			default:
				return translate( 'All languages', { textOnly: true } );
		}
	}

	getDisplayedLanguages() {
		const { languages } = this.props;
		const { search, filter } = this.state;

		if ( search ) {
			const searchString = search.toLowerCase();
			return languages.filter( language => {
				return includes( language.name.toLowerCase(), searchString );
			} );
		}

		switch ( filter ) {
			case 'popular':
				const popularLanguages = languages.filter( language => language.popular );
				popularLanguages.sort( ( a, b ) => a.popular - b.popular );
				return popularLanguages;
			default:
				return languages;
		}
	}

	handleSearch = ( search ) => {
		this.setState( { search } );
	}

	handleClick = ( selectedLanguageSlug ) => {
		this.setState( { selectedLanguageSlug } );
	}

	handleSelectLanguage = () => {
		const langSlug = this.state.selectedLanguageSlug;
		this.props.onSelected( langSlug );
		this.handleClose();
	}

	handleClose = () => {
		this.props.onClose();
	}

	renderTabItems() {
		const tabs = [ 'popular', '' ];

		return map( tabs, filter => {
			const selected = this.state.filter === filter;
			const onClick = () => this.setState( { filter } );

			return (
				<SectionNavTabItem
					key={ filter }
					selected={ selected }
					onClick={ onClick }
				>
					{ this.getFilterLabel( filter ) }
				</SectionNavTabItem>
			);
		} );
	}

	renderLanguageList() {
		const languages = this.getDisplayedLanguages();

		return (
			<div className="language-picker__modal-list">
				{ map( languages, this.renderLanguageItem ) }
			</div>
		);
	}

	renderLanguageItem = ( language ) => {
		const isSelected = language.langSlug === this.state.selectedLanguageSlug;
		const classes = classNames( 'language-picker__modal-text', {
			'is-selected': isSelected
		} );

		return (
			<div
				className="language-picker__modal-item"
				key={ language.langSlug }
				onClick={ partial( this.handleClick, language.langSlug ) }
			>
				<span className={ classes }>{ language.name }</span>
			</div>
		);
	}

	render() {
		const { isVisible, translate } = this.props;

		if ( ! isVisible ) {
			// Render nothing at all if the modal is not visible
			// <Dialog isVisible={ false }> still renders a lot of useless elements
			return null;
		}

		const buttons = [
			{
				action: 'cancel',
				label: translate( 'Cancel' )
			},
			{
				action: 'confirm',
				label: translate( 'Select Language' ),
				isPrimary: true,
				onClick: this.handleSelectLanguage
			},
		];

		return (
			<Dialog
				isVisible
				buttons={ buttons }
				onClose={ this.handleClose }
				additionalClassNames="language-picker__modal"
			>
				<SectionNav selectedText={ this.getFilterLabel( this.state.filter ) }>
					<SectionNavTabs>
						{ this.renderTabItems() }
					</SectionNavTabs>
					<Search
						pinned
						fitsContainer
						onSearch={ this.handleSearch }
						placeholder={ translate( 'Search languagesâ€¦' ) }
					/>
				</SectionNav>
				{ this.renderLanguageList() }
			</Dialog>
		);
	}
}

export default localize( LanguagePickerModal );
