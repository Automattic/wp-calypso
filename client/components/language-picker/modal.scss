.language-picker__modal {
	&.dialog.card {
		position: absolute;
		top: 0;
		left: 0;
		right: 0;
		bottom: 0;
		max-width: none;
		font-size: $font-body-small;
		display: flex;
		flex-direction: column;

		@include breakpoint-deprecated( '>660px' ) {
			top: 5%;
			bottom: 5%;
			left: 5%;
			right: 5%;
			width: 90%;
		}

		@include breakpoint-deprecated( '>960px' ) {
			left: 12.5%;
			right: 12.5%;
			width: 75%;
		}
	}

	.dialog__content {
		display: flex;
		flex-direction: column;
		flex: auto;
		min-height: 0; // permits the .language-picker__modal-list to shrink below its min height
		overflow: inherit;
		padding: 0;
	}

	.dialog__action-buttons {
		flex: none;
		margin: 0;
	}

	.section-nav {
		flex: none;
		margin-bottom: 0;
		z-index: 10;

		@include breakpoint-deprecated( '<660px' ) {
			.search.has-focus {
				margin: 0 4px;
				width: calc( 100% - 8px );
			}
		}
	}
}

.language-picker__modal-item {
	display: inline-block;
	white-space: nowrap;
	overflow: hidden;
	&:focus {
		outline: solid var( --color-neutral-light ) 1px;
	}
}

.language-picker__modal-item-notice {
	display: inline-block;
	vertical-align: middle;
	margin: -4px 0 -4px 4px;
	color: var( --color-warning );
	line-height: 1;
	visibility: hidden;

	.language-picker__modal-item:hover &,
	.language-picker__modal-item:focus &,
	.is-selected & {
		color: var( --color-warning-light );
		visibility: visible;
	}
}

.language-picker__modal-tooltip {
	z-index: z-index( 'root', 'popover.is-dialog-visible' );
}

.language-picker__modal-list {
	box-sizing: border-box;
	width: 100%;
	overflow-y: auto;
	padding: 8px 16px;
	flex: auto;

	.language-picker__modal-item {
		width: 100%;

		@include breakpoint-deprecated( '>480px' ) {
			width: 50%;
		}

		@include breakpoint-deprecated( '>660px' ) {
			width: 33%;
		}

		@include breakpoint-deprecated( '>960px' ) {
			width: 25%;
		}
	}
}

.language-picker__modal-text {
	display: inline-block;
	padding: 4px 8px;
	border-radius: 2px;
	cursor: pointer;

	&.is-selected {
		background: var( --color-primary );
		color: var( --color-text-inverted );
	}
}

.language-picker__modal-suggested {
	flex: none;
	background: var( --color-neutral-0 );
	border-top: 1px solid var( --color-neutral-0 );
	padding: 8px 16px;
	z-index: 1;
	position: relative;

	&::before {
		content: '';
		display: block;
		position: absolute;
		bottom: 100%;
		left: 16px;
		right: 16px;
		height: 24px;
		background: linear-gradient( to bottom, rgba( 255, 255, 255, 0 ) 0%, white 100% );
		margin-bottom: 1px;
	}

	.language-picker__modal-suggested-inner {
		display: flex;
		overflow: hidden;
	}

	.language-picker__modal-suggested-label {
		flex: none;
		padding: 4px 8px;
		white-space: nowrap;
	}

	.language-picker__modal-suggested-list {
		flex: 1 1 0;
		height: 0;
	}

	.language-picker__modal-suggested-list-inner {
		display: flex;
		flex-wrap: wrap;
	}

	.language-picker__modal-item {
		flex: none;
		width: auto;
		margin-left: 16px;
	}
}

.language-picker__modal-empathy-mode {
	display: flex;
	padding: 8px 16px;

	@include breakpoint-deprecated( '>480px' ) {
		justify-content: flex-end;
	}

	.form-label {
		position: relative;
		z-index: 1;
		margin: 0;
		padding: 4px 8px;
	}
}

.language-picker__modal-locale-notice {
	position: relative;
	z-index: 1;
	padding: 8px 16px;
	text-align: right;

	.form-label {
		display: inline-block;
		vertical-align: top;
		margin: 0;
	}

	p {
		margin: 0;
	}
}
