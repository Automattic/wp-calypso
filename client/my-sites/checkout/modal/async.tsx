import { useMemo } from 'react';
import AsyncLoad from 'calypso/components/async-load';
import { KEY_PRODUCTS } from './constants';
import type { Props } from '.';

const useProducts = () => {
	const { search } = window.location;
	const products = useMemo( () => new URLSearchParams( search ).get( KEY_PRODUCTS ), [ search ] );

	return products;
};

const AsyncCheckoutModal = ( props: Props ) => {
	const products = useProducts();

	if ( ! products ) {
		return null;
	}

	return (
		<AsyncLoad
			require="calypso/my-sites/checkout/modal"
			placeholder={ null }
			productAliasFromUrl={ products }
			{ ...props }
		/>
	);
};

export default AsyncCheckoutModal;
