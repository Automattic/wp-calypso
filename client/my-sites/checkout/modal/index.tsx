import { StripeHookProvider } from '@automattic/calypso-stripe';
import { Modal } from '@wordpress/components';
import { removeQueryArgs } from '@wordpress/url';
import { useTranslate } from 'i18n-calypso';
import page from 'page';
import { useSelector } from 'react-redux';
import CheckoutMasterbar from 'calypso/layout/masterbar/checkout';
import { getStripeConfiguration } from 'calypso/lib/store-transactions';
import CalypsoShoppingCartProvider from 'calypso/my-sites/checkout/calypso-shopping-cart-provider';
import CompositeCheckout from 'calypso/my-sites/checkout/composite-checkout/composite-checkout';
import getPreviousRoute from 'calypso/state/selectors/get-previous-route.js';
import isAtomicSite from 'calypso/state/selectors/is-site-automated-transfer';
import { isJetpackSite } from 'calypso/state/sites/selectors';
import { getSelectedSite, getSelectedSiteId } from 'calypso/state/ui/selectors';
import { KEY_PRODUCTS } from './constants';
import type { FunctionComponent } from 'react';

import './style.scss';

export interface Props {
	title?: string;
	productAliasFromUrl?: string;
	redirectTo?: string;
	checkoutOnSuccessCallback?: () => void;
	onClose?: () => void;
}

const CheckoutModal: FunctionComponent< Props > = ( {
	title = '',
	productAliasFromUrl,
	redirectTo,
	checkoutOnSuccessCallback,
	onClose,
} ) => {
	const translate = useTranslate();
	const { siteSlug, selectedSiteId, previousRoute, isJetpackNotAtomic } = useSelector(
		( state ) => {
			const site = getSelectedSite( state );
			const selectedSiteId = getSelectedSiteId( state );
			const previousRoute = getPreviousRoute( state );

			return {
				siteSlug: site?.slug,
				selectedSiteId,
				previousRoute: removeQueryArgs( previousRoute, KEY_PRODUCTS ),
				isJetpackNotAtomic:
					!! isJetpackSite( state, selectedSiteId ) && ! isAtomicSite( state, selectedSiteId ),
			};
		}
	);

	const handleRequestClose = () => {
		onClose?.();
		page( previousRoute );
	};

	const handleAfterPaymentComplete = () => {
		checkoutOnSuccessCallback?.();
		handleRequestClose();
	};

	return (
		<Modal
			open
			overlayClassName="checkout-modal"
			bodyOpenClassName="has-checkout-modal"
			title={ String( translate( 'Checkout modal' ) ) }
			shouldCloseOnClickOutside={ false }
			onRequestClose={ handleRequestClose }
		>
			<CheckoutMasterbar
				title={ title }
				siteSlug={ siteSlug }
				previousPath={ previousRoute }
				isJetpackNotAtomic={ isJetpackNotAtomic }
				isLeavingAllowed
			/>
			<CalypsoShoppingCartProvider>
				<StripeHookProvider
					fetchStripeConfiguration={ getStripeConfiguration }
					locale={ translate.localeSlug }
				>
					<CompositeCheckout
						siteId={ selectedSiteId ?? undefined }
						siteSlug={ siteSlug }
						productAliasFromUrl={ productAliasFromUrl }
						// Custom thank-you URL for payments that are processed after a redirect (eg: Paypal)
						redirectTo={ redirectTo || previousRoute }
						customizedPreviousPath={ previousRoute }
						isInModal
						disabledThankYouPage
						onAfterPaymentComplete={ handleAfterPaymentComplete }
					/>
				</StripeHookProvider>
			</CalypsoShoppingCartProvider>
		</Modal>
	);
};

export default CheckoutModal;
