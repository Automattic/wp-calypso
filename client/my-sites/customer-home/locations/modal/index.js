import VideoModal from 'calypso/components/videos-ui/video-modal';
import { COURSE_SLUGS } from 'calypso/data/courses';
import { useRouteModal } from 'calypso/lib/route-modal';
import PluginsAnnouncementModal from 'calypso/my-sites/plugins/plugins-announcement-modal';

const Modals = () => {
	const { isModalOpen, openModal, closeModal } = useRouteModal(
		'myHomeCoursePaymentsModal',
		COURSE_SLUGS.PAYMENTS_FEATURES
	);

	return (
		<>
			<PluginsAnnouncementModal />
			<VideoModal
				isVisible={ isModalOpen }
				onClose={ closeModal }
				onOpen={ openModal }
				courseSlug={ COURSE_SLUGS.PAYMENTS_FEATURES }
			/>
		</>
	);
};

export default Modals;
