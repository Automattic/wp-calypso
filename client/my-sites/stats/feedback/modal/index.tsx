import { Button, Modal, TextareaControl } from '@wordpress/components';
import { close } from '@wordpress/icons';
import { useTranslate } from 'i18n-calypso';
import React, { useState, useCallback, useEffect } from 'react';
import StatsButton from 'calypso/my-sites/stats/components/stats-button';
import useNoticeVisibilityMutation from 'calypso/my-sites/stats/hooks/use-notice-visibility-mutation';
import {
	NOTICES_KEY_ABLE_TO_SUBMIT_FEEDBACK,
	NOTICES_KEY_SHOW_FLOATING_USER_FEEDBACK_PANEL,
	useNoticeVisibilityQuery,
} from 'calypso/my-sites/stats/hooks/use-notice-visibility-query';
import { trackStatsAnalyticsEvent } from 'calypso/my-sites/stats/utils';
import { useDispatch } from 'calypso/state';
import { successNotice } from 'calypso/state/notices/actions';
import useSubmitProductFeedback from './use-submit-product-feedback';

import './style.scss';

interface ModalProps {
	siteId: number;
	onClose: () => void;
}

const FEEDBACK_SHOULD_SHOW_PANEL_API_KEY = NOTICES_KEY_SHOW_FLOATING_USER_FEEDBACK_PANEL;
const FEEDBACK_SHOULD_SHOW_PANEL_API_HIBERNATION_DELAY = 3600 * 24 * 30 * 12; // 12 months
const FEEDBACK_THROTTLE_SUBMISSION_DELAY = 60 * 5; // 5 minutes

const FeedbackModal: React.FC< ModalProps > = ( { siteId, onClose } ) => {
	const translate = useTranslate();
	const dispatch = useDispatch();
	const [ content, setContent ] = useState( '' );

	const {
		data: isAbleToSubmitFeedback,
		isFetching: isCheckingAbilityToSubmitFeedback,
		refetch: refetchNotices,
	} = useNoticeVisibilityQuery( siteId, NOTICES_KEY_ABLE_TO_SUBMIT_FEEDBACK );

	const { mutateAsync: disableFeedbackSubmission } = useNoticeVisibilityMutation(
		siteId,
		NOTICES_KEY_ABLE_TO_SUBMIT_FEEDBACK,
		'postponed',
		FEEDBACK_THROTTLE_SUBMISSION_DELAY
	);

	const { mutateAsync: updateFeedbackHibernationPeriod } = useNoticeVisibilityMutation(
		siteId,
		FEEDBACK_SHOULD_SHOW_PANEL_API_KEY,
		'postponed',
		FEEDBACK_SHOULD_SHOW_PANEL_API_HIBERNATION_DELAY
	);

	const { isSubmittingFeedback, submitFeedback, isSubmissionSuccessful } =
		useSubmitProductFeedback( siteId );

	const handleClose = useCallback( () => {
		setTimeout( () => {
			onClose();
		}, 200 );
	}, [ onClose ] );

	const onFormSubmit = useCallback( () => {
		if ( ! content ) {
			return;
		}

		trackStatsAnalyticsEvent( 'stats_feedback_action_submit_form', {
			feedback: content,
		} );

		const sourceUrl = `${ window.location.origin }${ window.location.pathname }`;
		submitFeedback( {
			source_url: sourceUrl,
			product_name: 'Jetpack Stats',
			feedback: content,
			is_testing: true,
		} );
	}, [ dispatch, content, submitFeedback ] );

	useEffect( () => {
		if ( isSubmissionSuccessful ) {
			dispatch(
				successNotice( translate( 'Thank you for your feedback!' ), {
					id: 'submit-product-feedback-success',
					duration: 5000,
				} )
			);

			updateFeedbackHibernationPeriod();
			disableFeedbackSubmission().then( () => {
				refetchNotices();
			} );

			handleClose();
		}
	}, [
		dispatch,
		isSubmissionSuccessful,
		handleClose,
		translate,
		disableFeedbackSubmission,
		updateFeedbackHibernationPeriod,
		refetchNotices,
	] );

	return (
		<Modal className="stats-feedback-modal" onRequestClose={ handleClose } __experimentalHideHeader>
			<Button
				className="stats-feedback-modal__close-button"
				onClick={ handleClose }
				icon={ close }
				label={ translate( 'Close' ) }
			/>
			<div className="stats-feedback-modal__wrapper">
				<h1 className="stats-feedback-modal__title">
					{ translate( 'Help us make Jetpack Stats better' ) }
				</h1>

				<div className="stats-feedback-modal__text">
					{ translate(
						'We value your opinion and would love to hear more about your experience. Please share any specific thoughts or suggestions you have to improve Jetpack Stats.'
					) }
				</div>
				<TextareaControl
					rows={ 5 }
					cols={ 40 }
					className="stats-feedback-modal__form"
					placeholder={ translate( 'Add your feedback here' ) }
					name="content"
					value={ content }
					onChange={ setContent }
					disabled={ ! isCheckingAbilityToSubmitFeedback && ! isAbleToSubmitFeedback }
				/>
				<div className="stats-feedback-modal__button">
					{ ! isCheckingAbilityToSubmitFeedback && ! isAbleToSubmitFeedback && (
						<strong>
							<em>
								{ translate( 'Feedback submission is currently limited to one per 5 minutes.' ) }
							</em>
						</strong>
					) }
					<StatsButton
						primary
						onClick={ onFormSubmit }
						busy={ isSubmittingFeedback }
						disabled={
							isCheckingAbilityToSubmitFeedback ||
							! isAbleToSubmitFeedback ||
							isSubmittingFeedback ||
							isSubmissionSuccessful ||
							! content
						}
					>
						{ translate( 'Submit' ) }
					</StatsButton>
				</div>
			</div>
		</Modal>
	);
};

export default FeedbackModal;
