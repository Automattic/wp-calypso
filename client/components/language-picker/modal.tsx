/**
 * External dependencies
 */
import { Dialog } from '@automattic/components';
import { useI18n } from '@automattic/react-i18n';
import { sprintf } from '@wordpress/i18n';
import type { I18nReact } from '@automattic/react-i18n';
import LanguagePicker, { createLanguageGroups } from '@automattic/language-picker';
import type { Language, LocalizedLanguageNames } from '@automattic/language-picker';
import React, { useState } from 'react';
import { connect } from 'react-redux';

/**
 * WordPress dependencies
 */
import { Button, Tooltip } from '@wordpress/components';
import { Icon, info } from '@wordpress/icons';

/**
 * Internal dependencies
 */
import QueryLanguageNames from 'calypso/components/data/query-language-names';
import FormLabel from 'calypso/components/forms/form-label';
import FormCheckbox from 'calypso/components/forms/form-checkbox';
import getLocalizedLanguageNames from 'calypso/state/selectors/get-localized-language-names';
import { isDefaultLocale, isTranslatedIncompletely } from 'calypso/lib/i18n-utils/utils';

/**
 * Style dependencies
 */
import './modal.scss';

type CalypsoLanguage = Language & {
	calypsoPercentTranslated: number;
	isTranslatedCompletely: boolean;
};

type Props = {
	languages: CalypsoLanguage[];
	/**
	 * Fires when the modal is dismissed.
	 */
	onClose: () => void;
	/**
	 * Fires when the modal is dismissed with a language selection.
	 */
	onSelectLanguage: (
		l: CalypsoLanguage,
		extraOptions: { empathyMode: boolean; useFallbackForIncompleteLanguages: boolean }
	) => void;
	selectedLanguageSlug?: string;
	localizedLanguageNames: LocalizedLanguageNames;
	empathyMode?: boolean;
	useFallbackForIncompleteLanguages?: boolean;
	showEmpathyModeControl?: boolean;
	getIncompleteLocaleNoticeMessage?: ( l: CalypsoLanguage ) => string;
};

type IncompleteLocaleNoticeMessageProps = {
	language: CalypsoLanguage | undefined;
	__: I18nReact[ '__' ];
	getIncompleteLocaleNoticeMessage: ( ( l: CalypsoLanguage ) => JSX.Element ) | undefined;
};

function IncompleteLocaleNoticeMessage( {
	language,
	__,
	getIncompleteLocaleNoticeMessage: getIncompleteLocaleNoticeMessageProp,
}: IncompleteLocaleNoticeMessageProps ): JSX.Element | null {
	if ( ! language ) {
		return null;
	}

	if ( typeof getIncompleteLocaleNoticeMessageProp === 'function' ) {
		return getIncompleteLocaleNoticeMessageProp( language );
	}

	const { name: languageName, calypsoPercentTranslated: percentTranslated } = language;
	const translateFAQLink = 'https://translate.wordpress.com/faq/';

	return (
		<Tooltip
			position="top center"
			text={
				<div className="language-picker__modal-incomplete-locale-tooltip-text">
					{ __( 'You can help translate WordPress.com into your language.' ) }
					<a href={ translateFAQLink }>{ __( 'Learn more' ) }</a>
				</div>
			}
		>
			<div className="language-picker__modal-incomplete-locale-notice-info">
				{
					/* translators: %(languageName)s is a localized language name, %(percentTranslated)d%% is a percentage number (0-100), followed by an escaped percent sign %%. */
					sprintf( __( '(%(languageName)s is only %(percentTranslated)d%% translated)' ), {
						languageName,
						percentTranslated,
					} )
				}
				<Icon icon={ info } size={ 20 } />
			</div>
		</Tooltip>
	);
}

const LanguagePickerModal: React.FC< Props > = ( {
	languages,
	onClose,
	onSelectLanguage,
	selectedLanguageSlug: selectedLanguageSlugProp,
	localizedLanguageNames,
	showEmpathyModeControl = false,
	empathyMode: initialEmapthyModeValue = false,
	useFallbackForIncompleteLanguages: initialUseFallbackForIncompleteLanguages = false,
	getIncompleteLocaleNoticeMessage: getIncompleteLocaleNoticeMessageProp,
} ) => {
	const { __ } = useI18n();
	const [ selectedLanguage, setSelectedLanguage ] = useState< CalypsoLanguage | undefined >(
		languages.find( ( l ) => l.langSlug === selectedLanguageSlugProp )
	);
	const [ empathyMode, setEmpathyMode ] = useState( initialEmapthyModeValue );
	const [ useFallbackForIncompleteLanguages, setUseFallbackForIncompleteLanguages ] = useState(
		initialUseFallbackForIncompleteLanguages
	);

	const selectedLanguageSlug = selectedLanguage && selectedLanguage.langSlug;

	const isDefaultLanguageSelected = selectedLanguageSlug && isDefaultLocale( selectedLanguageSlug );
	const showIncompleteLocaleControl =
		! isDefaultLanguageSelected &&
		selectedLanguageSlug &&
		isTranslatedIncompletely( selectedLanguageSlug );

	const showCheckboxes = showEmpathyModeControl || showIncompleteLocaleControl;
	const checkboxes = showCheckboxes ? (
		<div className="language-picker__modal-checkboxes">
			{ showIncompleteLocaleControl && (
				<div className="language-picker__modal-locale-notice">
					{ ! getIncompleteLocaleNoticeMessageProp && (
						<FormLabel>
							<FormCheckbox
								checked={ useFallbackForIncompleteLanguages }
								onChange={ ( e ) => setUseFallbackForIncompleteLanguages( e.target.checked ) }
							/>
							<span>
								<span className="language-picker__modal-incomplete-locale-text">
									{ __( 'Display interface in English' ) }
								</span>
								<span className="language-picker__modal-incomplete-locale-notice">
									<IncompleteLocaleNoticeMessage
										language={ selectedLanguage }
										__={ __ }
										getIncompleteLocaleNoticeMessage={ getIncompleteLocaleNoticeMessageProp }
									/>
								</span>
							</span>
						</FormLabel>
					) }
				</div>
			) }
			{ showEmpathyModeControl && (
				<div className="language-picker__modal-empathy-mode">
					<FormLabel>
						<FormCheckbox
							checked={ empathyMode && ! isDefaultLanguageSelected }
							disabled={ !! isDefaultLanguageSelected }
							onChange={ ( e ) => setEmpathyMode( e.target.checked ) }
						/>
						<span title="Pretend to use that language but display English where a translated exists">
							Empathy mode (a8c only)
						</span>
					</FormLabel>
				</div>
			) }
		</div>
	) : null;

	const buttons = [
		<>{ checkboxes }</>,
		<div className="language-picker__modal-buttons">
			<Button isLink onClick={ onClose }>
				{ __( 'Cancel' ) }
			</Button>
			<Button
				isSecondary
				onClick={ () => {
					onClose();
					if ( selectedLanguage ) {
						onSelectLanguage( selectedLanguage, {
							empathyMode,
							useFallbackForIncompleteLanguages,
						} );
					}
				} }
			>
				{ __( 'Apply Changes' ) }
			</Button>
		</div>,
	];

	return (
		<Dialog
			isVisible
			onClose={ onClose }
			buttons={ buttons }
			additionalClassNames="language-picker__dialog"
		>
			<QueryLanguageNames />
			<LanguagePicker
				headingTitle={ __( 'Select a language' ) }
				languages={ languages }
				languageGroups={ createLanguageGroups( __ ) }
				onSelectLanguage={ setSelectedLanguage }
				selectedLanguage={ selectedLanguage }
				localizedLanguageNames={ localizedLanguageNames }
			/>
		</Dialog>
	);
};

export default connect( ( state ) => ( {
	localizedLanguageNames: getLocalizedLanguageNames( state ) as LocalizedLanguageNames,
} ) )( LanguagePickerModal );
