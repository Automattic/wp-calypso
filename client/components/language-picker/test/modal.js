/**
 * @format
 * @jest-environment jsdom
 */

/**
 * External dependencies
 */
import React from 'react';
import { shallow } from 'enzyme';
import { identity, noop } from 'lodash';

/**
 * Internal dependencies
 */
import { LanguagePickerModal } from '../modal';

describe( 'LanguagePickerModal', () => {
	const defaultProps = {
		onSelected: noop,
		onClose: noop,
		isVisible: true,
		languages: [
			{
				langSlug: 'en',
				name: 'English',
				popular: 1,
				value: 1,
				wpLocale: 'en_US',
			},
			{
				langSlug: 'cs',
				name: 'Čeština',
				value: 11,
				wpLocale: 'cs_CZ',
			},
			{
				langSlug: 'it',
				name: 'Italiano',
				popular: 8,
				value: 35,
				wpLocale: 'it_IT',
			},
		],
		selected: 'en',
		translate: identity,
	};

	test( 'should render', () => {
		const wrapper = shallow( <LanguagePickerModal { ...defaultProps } /> );

		expect( wrapper ).toMatchSnapshot();
	} );

	describe( 'suggested languages', () => {
		const browserLanguages = [ 'en-AU', 'en', 'en-US', 'it' ];
		let _navigator;

		beforeAll( () => {
			_navigator = global.navigator;
			Object.defineProperty( global.navigator, 'languages', {
				get: () => browserLanguages,
			} );
		} );

		afterAll( () => {
			global.navigator = _navigator;
		} );

		test( 'should return browser languages that appear in WordPress.com supported languages', () => {
			const wrapper = shallow( <LanguagePickerModal { ...defaultProps } /> );

			expect( wrapper.instance().getSuggestedLanguages() ).toEqual( [
				defaultProps.languages[ 0 ],
				defaultProps.languages[ 2 ],
			] );
		} );

		test( 'should render a list of suggested languages', () => {
			const wrapper = shallow( <LanguagePickerModal { ...defaultProps } /> );
			const suggestedLanguagesTexts = wrapper.find(
				'.language-picker__modal-suggested-list .language-picker__modal-text'
			);

			expect( suggestedLanguagesTexts ).toHaveLength( 2 );
			expect( suggestedLanguagesTexts.at( 0 ).text() ).toEqual( 'English' );
			expect( suggestedLanguagesTexts.at( 1 ).text() ).toEqual( 'Italiano' );
		} );
	} );
} );
