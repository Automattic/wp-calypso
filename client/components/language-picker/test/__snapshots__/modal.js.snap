// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguagePickerModal should render 1`] = `
<Dialog
  additionalClassNames="language-picker__modal"
  buttons={
    Array [
      Object {
        "action": "cancel",
        "label": "Cancel",
      },
      Object {
        "action": "confirm",
        "isPrimary": true,
        "label": "Select Language",
        "onClick": [Function],
      },
    ]
  }
  isVisible={true}
  leaveTimeout={200}
  onClose={[Function]}
  onClosed={[Function]}
>
  <SectionNav
    allowDropdown={true}
    onMobileNavPanelOpen={[Function]}
    selectedText="Popular languages"
  >
    <NavTabs
      hasSiblingControls={false}
    >
      <NavItem
        key="popular"
        onClick={[Function]}
        selected={true}
      >
        Popular languages
      </NavItem>
      <NavItem
        onClick={[Function]}
        selected={false}
      >
        All languages
      </NavItem>
    </NavTabs>
    <Search
      autoFocus={false}
      compact={false}
      delaySearch={false}
      delayTimeout={300}
      disableAutocorrect={false}
      disabled={false}
      fitsContainer={true}
      hideClose={false}
      hideOpenIcon={false}
      isOpen={false}
      onClick={[Function]}
      onKeyDown={[Function]}
      onSearch={[Function]}
      onSearchChange={[Function]}
      onSearchClose={[Function]}
      onSearchOpen={[Function]}
      pinned={true}
      placeholder="Search languagesâ€¦"
      searching={false}
    />
  </SectionNav>
  <div
    className="language-picker__modal-list"
  >
    <div
      className="language-picker__modal-item"
      key="en"
      onClick={[Function]}
    >
      <span
        className="language-picker__modal-text is-selected"
      >
        English
      </span>
    </div>
    <div
      className="language-picker__modal-item"
      key="it"
      onClick={[Function]}
    >
      <span
        className="language-picker__modal-text"
      >
        Italiano
      </span>
    </div>
  </div>
  <div
    className="language-picker__modal-suggested"
  >
    <div
      className="language-picker__modal-suggested-inner"
    >
      <div
        className="language-picker__modal-suggested-label"
      >
        Suggested languages: 
      </div>
      <div
        className="language-picker__modal-suggested-list"
      >
        <div
          className="language-picker__modal-suggested-list-inner"
        >
          <div
            className="language-picker__modal-item"
            key="en"
            onClick={[Function]}
          >
            <span
              className="language-picker__modal-text is-selected"
            >
              English
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</Dialog>
`;
